### PR TITLE
[UNR-3626] 4.25 Support

### PR DESCRIPTION
**Pull request by mironec**
#### Description
4.25 things.
Associated engine PR: https://github.com/improbableio/UnrealEngine/pull/580.

#### Release note
Ah, I forgot.

#### Tests
A lot of testing, but will need a lot more.

#### Documentation
Will need the support matrix updated.

#### Primary reviewers
@jessicafalk @m-samiec @improbable-valentyn

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp [hunk 1]
```diff
@@ -193,8 +193,32 @@ void ComponentReader::ApplySchemaObject(Schema_Object* ComponentObject, UObject&
 
 			if (NetDriver->IsServer() || ConditionMap.IsRelevant(Parent.Condition))
 			{
+
+				// This is mostly copied from ReceivePropertyHelper in RepLayout.cpp
+				auto GetSwappedCmd = [&Cmd, &Cmds, &Parents, bIsAuthServer, &Replicator, &Channel, &Parent]() -> const FRepLayoutCmd&
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp:198]: Could this have been a re-usable function within the Engine?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h [hunk 2]
```diff
@@ -19,4 +19,9 @@ class USpatialPendingNetGame : public UPendingNetGame
 
 	virtual void SendJoin() override;
 
+#if ENGINE_MINOR_VERSION >= 25
+	//Made virtual in PendingNetGame.h, make it do nothing, as it would usually send a packet
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h:23]: // Made virtual in PendingNetGame.h, make it do nothing, as it would usually send a packet

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h [hunk 3]
```diff
@@ -19,4 +19,9 @@ class USpatialPendingNetGame : public UPendingNetGame
...
+	virtual void SendInitialJoin() override {}
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h:24]: I'm not even sure if we need this class anymore. This used to exist for client travel back when we had a different implementation of instansiating the spatial net driver.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 4]
```diff
@@ -338,11 +338,28 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
 		RepDataNode->Condition = Parent.Condition;
 		RepDataNode->RepNotifyCondition = Parent.RepNotifyCondition;
 		RepDataNode->ArrayIndex = PropertyNode->StaticArrayIndex;
+#if ENGINE_MINOR_VERSION >= 25
+		if (Class->IsChildOf(AActor::StaticClass()))
+		{
+			// Uses the same pattern as ComponentReader::ApplySchemaObject and ReceivePropertyHelper
+			if (UNLIKELY((int32)AActor::ENetFields_Private::RemoteRole == Cmd.ParentIndex))
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:345]: This is exactly the same. Refactor timeeeeeeeeeeeeeeeeeeeeeeeee?

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 5]
```diff
@@ -100,6 +100,7 @@ void FSpatialGDKEditorToolbarModule::StartupModule()
 
 	OnAutoStartLocalDeploymentChanged();
 
+#if ENGINE_MINOR_VERSION < 25
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp:103]: Where's my comments.

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,12 +9,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [`x.y.z`] - Unreleased
 
+### Breaking changes:
+- Unreal Engine version `4.23` is no longer supported. We recommend upgrading to the newest version (`4.25.1`) to continue receiving updates.
```
Comment (joshuahuburn) [inline CHANGELOG.md:13]: 4.25.2????????????????????????????????????

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,12 +9,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- When upgrading to Unreal Engine 4.25 you must:
+  1. In the engine folder, run `git fetch && git checkout 4.25-SpatialOSUnrealGDK-0.11.0`
+  1. Download and install the -v16 clang-9.0.1-based toolchain from [this Unreal Engine Documentation page](https://docs.unrealengine.com/en-US/Platforms/Linux/GettingStarted/index.html).
```
Comment (joshuahuburn) [inline CHANGELOG.md:16]: Is this now in the ci image? `clang-9.0.1`

Comment (joshuahuburn) [reply CHANGELOG.md:16]: Make a follow-up if not already done.

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,12 +9,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  1. Run `Setup.bat`, which is located in the root directory of the UnrealEngine repository.
```
Comment (joshuahuburn) [inline CHANGELOG.md:17]: I believe `InstallGDK.bat` runs these batch files for you now. Might be worth looking into.

File: CHANGELOG.md [hunk 9]
```diff
@@ -9,12 +9,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
 ### Features:
 - You can now change the GDK Editor Setting `Stop local deployment on stop play in editor` in order to automatically stop deployment when you stop playing in editor.
 - Added the `Connect local server worker to the cloud deployment` checkbox in **SpatialOS Editor Settings**, that enables/disables the option to start and connect a local server to the cloud deployment when `Connect to cloud deployment` is enabled.
 - Added the ability to suppress RPC warnings of the form "Executed RPC <RPCName> with unresolved references" by RPC Type using new SpatialGDKSetting RPCTypeAllowUnresolvedParamMap.
 - Decoupled QueuedIncomingRPCWaitTime from reprocessing flush time with new parameter QueuedIncomingRPCRetryTime (default value 1.0s).  This enables independent control over how long to wait for queued RPCs to resolve parameters, as well as how frequently to check if the parameters are resolved.
 - Command-line arguments are now only available in non-shipping builds, if you wish to use command-line arguments for shipping builds the target rule `bEnableSpatialCmdlineInShipping` will let you do so.
+- Unreal Engine version `4.25.1` is now supported.
```
Comment (joshuahuburn) [inline CHANGELOG.md:28]: - Unreal Engine version `4.25.1` is now supported. Good luck.

File: CHANGELOG.md (hunk 8 above)
Comment (mironec) [reply CHANGELOG.md:17]: Checked now, don't think this is the case yet (we generateProjectFiles, but not Setup.bat)

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialPendingNetGame.h:24]: Removed my modification here, as it does not seem needed anymore, created ticket to remove SpatialPendingNetGame entirely: https://improbableio.atlassian.net/browse/UNR-3920.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:345]: Took a further look, and unfortunately this is not exactly the same, as the actions inside are different.
Could still be refactored with a function taking two lambdas, but I feel like that could be over-complicating it, and Epic also have a couple variations of this code.

File: CHANGELOG.md (hunk 7 above)
Comment (mironec) [reply CHANGELOG.md:16]: https://improbableio.atlassian.net/browse/UNR-3921

File: CHANGELOG.md [hunk 10]
```diff
@@ -9,14 +9,25 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [`x.y.z`] - Unreleased
 
+### Breaking changes:
+- Unreal Engine version `4.23` is no longer supported. We recommend upgrading to the newest version (`4.25.2`) to continue receiving updates.
```
Comment (m-samiec) [inline CHANGELOG.md:13]: I see we haven't removed the <=4.23 changes though

File: CHANGELOG.md (hunk 9 above)
Comment (m-samiec) [reply CHANGELOG.md]: And good night

**Comment by m-samiec** [comment]
> A lot of testing, but will need a lot more.

Have we updated the example project or test gyms? Would be good to give the NFRs a run asap.

**Comment by mironec** [comment]
TestGyms and Example project did not need updating, only some deprecation warning fixes that Valentyn already merged in, I think, and they are #if-guarded.

File: ci/unreal-engine.version [hunk 11]
```diff
@@ -1 +1,2 @@
+HEAD feature/4.25-support
```
Comment (joshuahuburn) [inline ci/unreal-engine.version:1]: I remember something being related to ordering of these lines for EP building. I assume the top one is the one used for EP?

Comment (mironec) [reply ci/unreal-engine.version:1]: Yeah, will need to be updated to 4.25-SpatialOSUnrealGDK anyway.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 12]
```diff
@@ -26,7 +26,7 @@
 #include "Utils/SpatialActorUtils.h"
 #include "Utils/SpatialDebugger.h"
 
-#include "Engine.h"
+#include "Engine/Engine.h"
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:29]: Drive-by include fix

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialStatics.h [hunk 13]
```diff
@@ -22,14 +22,14 @@ struct FLockingToken
 {
 	GENERATED_USTRUCT_BODY()
 
-    UPROPERTY(BlueprintReadOnly)
+	UPROPERTY(BlueprintReadOnly, Category = "SpatialGDK|Locking")
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialStatics.h:25]: Drive-by compilation fix. Not sure under what circumstances it happens, but Engine vs Project plugin, maybe?

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlagsTest.cpp [hunk 14]
```diff
@@ -4,7 +4,7 @@
 
 #include "Tests/TestDefinitions.h"
 #include "Tests/AutomationCommon.h"
-#include "Runtime/EngineSettings/Public/EngineSettings.h"
+#include "Runtime/EngineSettings/Classes/GeneralProjectSettings.h"
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/Misc/SpatialActivationFlagsTest.cpp:7]: Drive-by include fix
